docs/Specifications/Workgroup Creation & Management/Interfaces and schemas for the main logical input objects and resolvers.md

```diff
@@ -0,0 +1,233 @@
+
+## Keystore resolver interface
+
+The keystore resolver defines an abstract interface to be implemented. This interface governs how keystores are to be inputted in the baseline tooling and what the expected outputs are.
+
+Any keystore resolver should be a class that adheres to the following interface:
+
+    interface KeystoreResolver {
+    	resolve(extraParams) : Promise<ethers.Signer>
+    }
+
+Invocation of the `resolve` method of the resolver should return a promise returning an object adhering to the `ethers.Signer` ([https://github.com/ethers-io/ethers.js/blob/master/src.ts/abstract-signer.ts#L12](https://github.com/ethers-io/ethers.js/blob/master/src.ts/abstract-signer.ts#L12)). In the future, we might decide to change to even more generic signer or web3 instance.
+
+The `extraParams` are additional resolver-specific customization params passed as an input that might be needed for the resolver to work. The idea behind `extraParams` is to allow for further customizations of the resolver behavior. An example of extra params might be headers objects for a REST API based resolver.
+
+## Inbuilt Baseline Keystore resolvers
+
+### KeystorePathResolver
+
+As part of the Radish34, a `KeystorePathResolver` is to be developed. This resolver would adhere to the `KeystoreResolver` interface.
+
+    class KeystorePathResolver {
+    	constructor(keystorePath) { ... }
+    	resolve(extraParams) { ... }
+    }
+
+The `keystorePath` param is resolver specific parameter to allow the resolver to find and resolve the input data to the resulting object. It needs to be a system path to a file. Currently, this file needs to adhere to the following schema:
+
+    {
+    	signingKey: {
+    		address: string,
+    		mnemonic: string,
+    		privateKey: string
+    	}
+    }
+
+The resolver parses this object and creates `ethers.Signer` object out of it.
+
+The `extraParams` param is unused for this resolver.
+
+## Example of a custom resolvers
+
+An example of a custom `KeystoreResolver` that the community might find valuable is a Keystore  backed by a cloud HSM. HSM protected keys are the highest state of security, something that the highest standards companies would want/need to adhere to in order to ensure maximum security of their systems. As long as this resolver returns an object adhering to the `ethers.Signer` interface, the baseline tooling will work correctly. In the meantime, the `sendTransaction` and `signMessage` methods of the resulting objects can be leveraging the HSM infrastructure.
+
+# Contract Artifacts
+
+## Contract artifacts as logical input
+
+The contract artifacts are an object containing the necessary data to deploy or interact with a certain contract. In order to be used by the baseline tooling, the resolved contract object needs to follow the following schema:
+
+`ContractArtifactsSchema` :
```

Comment:
Can we require the contract artifact schema to include the maximum output from available `solc` compiler options? IIRC the schema here is just missing `assembly` and `opcodes`. It would also be nice to include the raw `source` here as well.

Reply:
I've looked this from the perspective of what is needed for the contract deployment to happen. What ways do you want to use the assembly and opcodes?

Also maybe we can define this as the "minimum" schema.

Reply:
The `opcodes` are needed to fingerprint contract-internal transactions and are pretty critical to optimal UX. i.e., if a contract deploys another contract, fingerprinting the contract-internal opcodes allow us to resolve the ABI(s) of such contracts deployed as a result of the contract-internal `CREATE` opcode. There are more use-cases beyond this but this one is critical.